nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-processors/src/main/java/org/apache/nifi/processors/kafka/KafkaPublisher.java

```diff
@@ -78,152 +81,182 @@
-     * of each content stream segment that had failed to be sent to Kafka, so
-     * upon retry only the failed segments are sent.
+     * multiple Kafka messages. The
+     * {@link PublishingContext#getLastAckedMessageIndex()} will provide the
+     * index of the last ACKed message, so upon retry only messages with the
+     * higher index are sent.
      *
-     * @param messageContext
-     *            instance of {@link SplittableMessageContext} which hold
-     *            context information about the message to be sent
-     * @param contentStream
-     *            instance of open {@link InputStream} carrying the content of
-     *            the message(s) to be send to Kafka
-     * @param partitionKey
-     *            the value of the partition key. Only relevant is user wishes
-     *            to provide a custom partition key instead of relying on
-     *            variety of provided {@link Partitioner}(s)
-     * @param maxBufferSize maximum message size
-     * @return The set containing the failed segment indexes for messages that
-     *         failed to be sent to Kafka.
+     * @param publishingContext
+     *            instance of {@link PublishingContext} which hold context
+     *            information about the message(s) to be sent.
+     * @return The index of the last successful offset.
      */
-    BitSet publish(SplittableMessageContext messageContext, InputStream contentStream, Integer partitionKey,
-            int maxBufferSize) {
-        List<Future<RecordMetadata>> sendFutures = this.split(messageContext, contentStream, partitionKey, maxBufferSize);
-        return this.publish(sendFutures);
+    KafkaPublisherResult publish(PublishingContext publishingContext) {
+        StreamDemarcator streamTokenizer = new StreamDemarcator(publishingContext.getContentStream(),
+                publishingContext.getDelimiterBytes(), publishingContext.getMaxRequestSize());
+
+        int prevLastAckedMessageIndex = publishingContext.getLastAckedMessageIndex();
+        List<Future<RecordMetadata>> resultFutures = new ArrayList<>();
+
+        byte[] messageBytes;
+        int tokenCounter = 0;
+        for (; (messageBytes = streamTokenizer.nextToken()) != null; tokenCounter++) {
+            if (prevLastAckedMessageIndex < tokenCounter) {
+                Integer partitionId = publishingContext.getPartitionId();
+                if (partitionId == null && publishingContext.getKeyBytes() != null) {
+                    partitionId = this.getPartition(publishingContext.getKeyBytes(), publishingContext.getTopic());
+                }
+                ProducerRecord<byte[], byte[]> message =
+                        new ProducerRecord<>(publishingContext.getTopic(), publishingContext.getPartitionId(), publishingContext.getKeyBytes(), messageBytes);
+                resultFutures.add(this.kafkaProducer.send(message));
+            }
+        }
+
+        tokenCounter -= 1;
```

Comment:
this decrement then increment appears to be left over from some previous or incomplete logic.  Probably can remove this line and drop the increment that follows, right?